### PR TITLE
Always generate a card

### PR DESCRIPTION
**Pull request by Arthur-Milchior**
This partially port a change from Anki 2.1.28. Notes with no card are not deleted, instead first card is generated. This also change import of decks. It does not change when cards should be generated from the add card window (this would require a change in UI, so we would need to discuss what we want to do exactly to port this change)

This ensure that "Empty cards" does not delete note with no cards, so no data is lost. 

# Test:

create a basic and a cloze card.

Edit them to remove front and cloze 1.

Click on "empty card"

See that it does not offer any deletion (i.e. no deletion of note, as in upstream)

### CODE REVIEW

**Comment by Arthur-Milchior** [comment]
I added a regression test for a bug related to cloze type with no actual cloze, whose first card were not detected as empty

**Comment by Arthur-Milchior** [comment]
Conflict removed

**Comment by mikehardy** [comment]
Took another batch of conflicts, sorry @Arthur-Milchior - on the plus side I did just merge something more than 10 PRs so we're moving along quite a bit

**Comment by mikehardy** [comment]
I can't believe this took a conflict? I mean, who is changing the card generation code :thinking: :rofl:

**Comment by mikehardy** [comment]
Just took a batch of conflicts unfortunately, but we are down to just a couple PRs in flight, so this might be the last round of de-conflicting...

**Comment by Arthur-Milchior** [comment]
I'd like to note that this PR can probably be reviewed even with merge conflict. It's not a NF change, so just deciding whether this is actually a change we want seems relevant. In particular, what to do when someone tries to generate a note which would not actually generate a card does not necessarily have to follow upstream given that we can't as easily go from "add card" to the browser correct the card

**Comment by mikehardy** [comment]
I'm okay with following upstream as closely as possible. I don't think this will be a common case at all so just doing what upstream does as closely as possible will at least align our behavior with any user help available (docs or posts online etc) for them to help themselves I think

**Review by mikehardy — APPROVED** [review]
Looks fine to me assuming tests pass

**Comment by Arthur-Milchior** [comment]
@david-allison-1 this requires your approval please. This would be nice to have it soon, since it allows to avoid data loss, and that's a huge difference with anki

**Comment by david-allison-1** [comment]
@Arthur-Milchior Could you list the upstream changes that this follows (ideally in the commit messages as well as the outputs) so I can compare?

**Comment by david-allison-1** [comment]
I also think that "Check Database" will require a change here, but haven't verified.

**Comment by Arthur-Milchior** [comment]
I believe that https://github.com/ankitects/anki/commit/f637ac957d206ead619d54d6cf86eb8958ea508c is a part of the commit you want. However, this commit just move code from py to rust. The code seems to have been added in https://github.com/ankitects/anki/commit/f24dc05c8d3a9c3b0b20f6f18e793d27bccef744
But I don't speak rust so I can't really check.

In this case, I believe that the change log of 2.1.28 and just reproducing the examples will help more 

I don't see anyplace in check database that check whether a note has a card, so I don't expect it to be useful.

**Comment by mikehardy** [comment]
If @david-allison-1 is okay with it, let's merge and move on - it looked fine to me in a scan a while back, and this is the time in 2.15 to merge items like this IMHO

**Comment by david-allison-1** [comment]
I'm going to need to go through this (Anki Desktop's behaviour) with a finer toothcomb, especially listing the commits which we're emulating here.

* What happens when previewing a note which generates no cards
* Could we create (or link to) unit tests which ensure that we've specified what Anki Desktop's new behaviour is, and that we match it.

**Comment by david-allison-1** [comment]
@Arthur-Milchior 

The relevant changes are 

* https://github.com/ankitects/anki/commit/f86c2dc567b321f9e5e42d0575ad3f834da7ec6e
* https://github.com/ankitects/anki/commit/15ff279a96f02ca3156ff0fed42efd8f3dab416d
* https://github.com/ankitects/anki/commit/bee0eb126426ad5cef9ca508ed1c9eecb9024fb8

These should provide enough context for tests

**Comment by Arthur-Milchior** [comment]
> These should provide enough context for tests

Only if you can understand this code. I've quite few intuition when I read Rust code and would not know how it does things.

How do you preview a note ? I only know how to preview some cards.
If I create a card which generate no note (e.g. inversing Front and Back in basic, and filling only "Back" side), it generate the card. If I open the browser and try to preview it, it simply shows:
"The front of this card is blank.
More information"

This is already what we have in ankidroid, since we check whether a card is empty before showing it.

When you mention linking unit test, are you asking to uses current tests in rust for the java implementation? No idea how you would do that, but you're the rust expert here.


Do you believe you'll have time to port upstream test if you understand them?
I should warn you that 2.1.28 made other change to card generation rules. In particular 
```
{{#A}}
  {{B}}
{{/A}}
{{^A}}
  {{C}}
{{/A}}
```
now correctly uses the rule "generate if (A and B) or (Not A and C)", but in ankidroid, we still don't accept rules as complex. It was my opinion that dealing with the change of card generation rules should be another distinct PR. If you try to port tests for such template, it is normal that they'll fail.

File: AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java [hunk 1]
```diff
@@ -892,7 +893,8 @@ public void _remNotes(java.util.Collection<Long> ids) {
 	    ArrayList<JSONObject> cms;
 	    switch (type) {
             case ADD:
-    	        cms = findTemplates(note);
+                // Don't show empty card, so it's clear during preview that no cards are generated
+    	        cms = findTemplates(note, false);
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java:897]: This code isn't used, what is the intention when a user is previewing?

File: AnkiDroid/src/main/java/com/ichi2/libanki/Models.java [hunk 2]
```diff
@@ -1111,16 +1110,21 @@ public static boolean emptyStandardCard(String type, JSONArray req, String[] tri
     /**
      * @param m A model
      * @param sfld Fields of a note
+     * @param allowEmpty whether to always return an ord, even if all cards are actually empty
      * @return The index of the cards that are generated. For cloze cards, if no card is generated, then {0} */
-    public static ArrayList<Integer> availOrds(Model m, String[] sfld) {
+    public static ArrayList<Integer> availOrds(Model m, String[] sfld, boolean allowEmpty) {
         if (m.getInt("type") == Consts.MODEL_CLOZE) {
-            return _availClozeOrds(m, sfld);
+            return _availClozeOrds(m, sfld, allowEmpty);
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Models.java:1117]: We've moved from `_availClozeOrds(.., true)` to `_availClozeOrds(.., variable)`, is this correct? Can this ever be false in Anki Desktop?

**Comment by david-allison-1** [comment]
Aside: I'll be putting up the Rust repo later on today as I was hitting my head against a brick wall trying to get the CI running locally.

TL;DR: We're making a major functionality change, I'm concerned that the code we're porting wasn't linked and there are no tests over how "empty cards" has changed. Please push back on any of the tests that you feel are a waste of time, or don't add value, I fear I've gone overboard, but don't see where I'd cut down.

I think we either have a bug, or this needs more testing and specification, as d2d0c078d8d4e0121b6b2758b8803ba62dc57075 should block cloze note types being added with no deletions (maybe we need to discuss and revisit this).

----

> How do you preview a note ? I only know how to preview some cards.

`Open Note Editor - Press Preview`. We need to know what behaviour that we will provide for this feature when generating an "empty" card. Maybe we need to explain this to the user better as previewing while creating/editing is not nearly used as often in Anki Desktop (as it comes from the Cards...) menu. 

We need to test this and ensure that it doesn't crash, and the messages are sensible, both for standard and cloze templates.

> When you mention linking unit test, are you asking to uses current tests in rust for the java implementation?

I mean providing a link to the current tests we have that exercise this functionality (if they exist), as justification that things are still working correctly/do not need testing.

We need to specify the behaviour that we're expecting from these changes, and put in both some low-level unit tests (to ensure that the Java does as expected) and high-level UI/Integration tests which specify the new functionality from both an "add card" and "empty cards" functionality point of view which won't change when we change the implementations over.

To do this, we need to understand the context of the change. We're a port, and we should look at (or use) the code that we're porting before making a change.

The implementation of the Java functionality is guaranteed to change when we move to the Rust backend, but we need to ensure that this change is correct if we get it in first. Architecturally, we should aim to ensure that the functionality can be replaced with a shim which would either call our Java implementation, or a Rust implementation defined in https://github.com/ankitects/anki/blob/master/rslib/backend.proto


> Do you believe you'll have time to port upstream test if you understand them?

No, and we shouldn't, we'll be using the Rust code, therefore we'll be running the Rust based unit tests. 

----

From a brief skim:

* Identify the relevant functions in: https://github.com/ankitects/anki/blob/master/rslib/backend.proto
* Quickly state that they'd be easy to port the code to, if not then either change our implementation, or explain why and leave it as a "we'll cross this bridge when we come to it"

For testing:

**Low-Level**

"Does the Java work as I'd expect it to work": we can probably get away with manual work here


**High Level**

Activity-based - test the workflows end-to-end, aim is to ensure that we match Anki Desktop, these are the tests that won't change after the shim is replaced with the Rust.

* Generate a card template which will not produce a note
   * Test that it can be added
      * Will there be a warning displayed? Does Anki Desktop? Will we get more users asking "Why does my card say that I need to add fields/cloze deletions", rather than "Why was my "Add" rejected with `This note type must contain cloze deletions`)).
   * Test the properties on the added card, especially `ord, due, did` as these are specified in the Rust code
   * Preview, ensure there is no crash
      * One card is available
      * Check the card content and that it's sensible and reasonably matches the Anki Desktop message
* Generate a cloze note with no deletions
   * Ditto, but check for different message on the preview.
* Test the creation of a cloze note with `{{c501::ABC}}` - ensure this matches Anki Desktop behaviour
* Empty Cards
    * Ensure that the "Keep notes with no valid cards" checkbox is displayed - do we do this? Aside: Has Anki Desktop implemented this yet?
    * Ensure behaviour both with and without the checkbox matches expectations

* Make Design Decision: Will this change use HTML for the report as the Rust does, or do we save this as a "bridge to cross later" for V2/3 of the rust conversion? 

Additional question: Are these cards scheduled as normal cards, can they be answered?

File: AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java (hunk 1 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java:897]: I just answered in https://github.com/ankidroid/Anki-Android/pull/7842 . In particular I explained what the method does (since there is no comment). I am confident that this change is consistent with the function goal. However, since I can't imagine we'll need this function, as we currently have a working previewer, I suggest to delete the method instead of finding the correct change to do

File: AnkiDroid/src/main/java/com/ichi2/libanki/Models.java (hunk 2 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/Models.java:1117]: I believe it is correct. As the PR state, my goal is to change a single thing: a note will always generate a card. In particular this will ensure that using "Empty card" and "Check Database" does not risk to delete cards by accident.

However, there remains plenty of differences that I did not port between 2.1.27 and 2.1.28. For example, anki now allows to add notes even when they does not generate a card. It shows a warning but register the note correctly. It is not clear to me why it does that, but it does. I have read that this feature was asked for a usecase I can't remember. My goal was not to port this change to AnkiDroid. First because if we do so, it should be in a separate PR I believe. Secondly because this would be a disaster until the UI is changed. Indeed, AnkiDroid does not have a way to access recently created card the way than anki does with the "History" button. Accepting note without cards is okay in Anki, because you can easily go back and edit it. It would not be okay in ankidroid without a way to edit the card recently added.

This mean that currently, we want "allow empty" to be `false` when we create a note (so that we don't allow to create notes that generate no card), but we want it to be `true` when using "Empty cards" in order to ensure we don't delete notes by accident. This is note something that exists upstream because they changed the way they deal with creation of card and my goal was not to change it.

**Comment by Arthur-Milchior** [comment]
> TL;DR: We're making a major functionality change

I am extremely confused. I would never have thought that "don't delete a note in the collection if it has no more card" would be called a major functionality change. Normally, it would not affect everyday workflow and just avoid losing data in some erroneous case.

> , I'm concerned that the code we're porting wasn't linked and there are no tests over how "empty cards" has changed.

Do you mean no test upstream, or not test here?
If I recall correctly, DAE explained the test are directly in rust, and so rejected when I wanted to add test in python. I don't understand the code of the test, but they exists and can probably be ported.

> I think we either have a bug, or this needs more testing and specification, as [d2d0c07](https://github.com/ankidroid/Anki-Android/commit/d2d0c078d8d4e0121b6b2758b8803ba62dc57075) should block cloze note types being added with no deletions (maybe we need to discuss and revisit this).

I entirely agree with you, it's far better not to accept to add notes that don't generate card. It's really nice that we differ from upstream here because the UX does not allow to edit a card as easily.
I was not suggested to change anything about creation of new note. Actually, if you run this code, a creation of a cloze type without cloze is still rejected.

I personally don't feel the need to revisit this choice. Unless we hear any complaints about it, I believe it's the right one.

> 
> https://github.com/ankidroid/Anki-Android/blob/1bae96da9faaa063e39c6a4da10c3caf7ecaf73d/AnkiDroid/src/test/java/com/ichi2/anki/NoteEditorTest.java#L165-L175
> 
> > How do you preview a note ? I only know how to preview some cards.
> 
> `Open Note Editor - Press Preview`. We need to know what behaviour that we will provide for this feature when generating an "empty" card. Maybe we need to explain this to the user better as previewing while creating/editing is not nearly used as often in Anki Desktop (as it comes from the Cards...) menu.
> 
> We need to test this and ensure that it doesn't crash, and the messages are sensible, both for standard and cloze templates.

So, I'm assuming we are discussing `NoteEditor`'s call to `performPreview`. If so, what it does is sending the note type and the field values to `CardTemplatePreviewer` (the name confuses me, since it seems not related to card template but to card directly).

In this case, it does not try to generate any cards. Instead, it only shows the question of the card that was selected when the note editor was opened. Since card generation never occurs and I don't change the rules about whether a single card is empty or not, this PR should not lead to any difference.

> 
> > When you mention linking unit test, are you asking to uses current tests in rust for the java implementation?
> 
> I mean providing a link to the current tests we have that exercise this functionality (if they exist), as justification that things are still working correctly/do not need testing.

I believe I added the test for the single change that this PR is supposed to make.
The test you suggested above is nice, and should be added to the test base independtly of this PR, since this it's supposed to be as true with and without this PR

> We need to specify the behaviour that we're expecting from these changes, and put in both some low-level unit tests (to ensure that the Java does as expected) and high-level UI/Integration tests which specify the new functionality from both an "add card" and "empty cards" functionality point of view which won't change when we change the implementations over.

Same remark. I am totally okay with those tests, and we can already incorporate it. I just don't see how they relate to this PR
 
> > Do you believe you'll have time to port upstream test if you understand them?
> 
> No, and we shouldn't, we'll be using the Rust code, therefore we'll be running the Rust based unit tests.

You lost me. If you want to add more unit tests to be sure we are similar to upstream, what unit test can we add if we don't add a copy of upstream's unit test? 

> From a brief skim:
> 
> * Identify the relevant functions in: https://github.com/ankitects/anki/blob/master/rslib/backend.proto

`GetEmptyCards` . So we want `fn get_empty_cards` in https://github.com/ankitects/anki/blob/master/rslib/src/backend/mod.rs and then all of https://github.com/ankitects/anki/blob/master/rslib/src/notetype/emptycards.rs

The important part is only whether `will_delete_note` is set or not.

Then the actual change is made in https://github.com/ankitects/anki/blob/master/qt/aqt/emptycards.py where, if the user did not click on the button to actually delete notes, the method `_delet_cards` ensure that the first card is not deleted

> * Quickly state that they'd be easy to port the code to, if not then either change our implementation, or explain why and leave it as a "we'll cross this bridge when we come to it"

We could, I guess, port the whole `emptycards.py` and `emptycards.rs`. But that would be a lot of work. As I stated, the rules for card generation changed a lot. It was not my goal currently to port all changes at once, but only a simple change: ensure each note in the database has at least one card.


> For testing:
> 
> **Low-Level**
> 
> "Does the Java work as I'd expect it to work": we can probably get away with manual work here

Do you mean to test the change introduced by this PR ? If so, I already do it. Else, what exactly do you want to test. There are a lot of things that could be interseting to test, but that are not relevant to this PR.

> 
> **High Level**
> 
> Activity-based - test the workflows end-to-end, aim is to ensure that we match Anki Desktop, these are the tests that won't change after the shim is replaced with the Rust.
> 
> * Generate a card template which will not produce a note
>   
>   * Test that it can be added

We never did any test to check whether a card template is valid or not. This was added recently in Anki, and I assume it would be nice to port it. E.g. it nows refuse to save a note type with a template `{{foo` or `{{#foo}}{{bar}}` But that does not seems to relate to this PR at all. I'm dealing with card and not with card type here.

>     * Will there be a warning displayed? Does Anki Desktop? Will we get more users asking "Why does my card say that I need to add fields/cloze deletions", rather than "Why was my "Add" rejected with `This note type must contain cloze deletions`)).

Currently, when you use "Empty cards", Anki gives a lot of details. See my collection e.g.:
2020-12-08-073648_1276x1401_scrot 
For a simple example with a note with all card empty:
2020-12-08-073741_531x345_scrot
The warning is shown in red. 

However, adding a detailled list of all cards that will be deleted seems hard to do in our UX. In particular, Anki gives link that will open the browser, and for each note we can check it's field and get more information before accepting or rejecting to delete. We probably don't want to copy this UX at all. If we decide to give as many details as Anki, that would be really cool, but we must discuss exactly which UX we want to show users, and once again I really fell to see how it's relevant to this PR.

At most, I would understand if you want to ask a warning "Some note have no more card anymore. Do you want to delete the notes or to keep an empty card to correct the note later", or something like that.

> Additional question: Are these cards scheduled as normal cards, can they be answered?

Those are indeed normal card. Imagine that you take a card "Basic (Optional reversed card)", that you fill all fields and then later that you remove the "Add reverse" content.
You'll get Card2, which is in your collection which the reviewer may still try to show you, but where it will tell you that the card is empty.  There is no difference between this card and the card that would be generated by "Empty cards" if by accident a note would have no more card anymore.

**Comment by Arthur-Milchior** [comment]
I'll try to write down a use case where this case is extremely important in my opinion.

Let assume that you have your collection filled with Basic cards. At some point, you try to understand what card template is, and you make a typo, you replace `{{Front}}` by `{Front}`

Now, all your notes have only empty card. If you use "Empty cards", you'll delete your whole collection. Thanks to this change, you'll lose nothing and be able to correct your card template.

Of course, this is a best case scenario. If instead you have "Basic and reverse" and that you make the same mistake in front and in back, then "Empty cards" will remove all back cards and keep front cards. That's clearly not ideal, but at least you only loose card, you don't loose your notes.

**Comment by mikehardy** [comment]
This is the exact sort of change I'd like to see in with a track record of alpha usage, so I'd like to see this get in shape and merge this sooner rather than later
@Arthur-Milchior - looks like there are some conflicts to resolve?
@david-allison-1 have your concerns been addressed? Last comments were from Arthur and there was a push after

File: AnkiDroid/src/main/java/com/ichi2/libanki/importer/NoteImporter.java [hunk 3]
```diff
@@ -376,7 +376,7 @@ private boolean processFields(ForeignNote note, @Nullable String[] fields) {
             }
         }
         note.fieldsStr = joinFields(fields);
-        ArrayList<Integer> ords = Models.availOrds(mModel, fields);
+        ArrayList<Integer> ords = Models.availOrds(mModel, fields, true);
         if (ords.isEmpty()) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/libanki/importer/NoteImporter.java:380]: This should always be false now

**Review by david-allison-1 — APPROVED** [review]
LGTM. One minor comment

File: AnkiDroid/src/main/java/com/ichi2/libanki/importer/NoteImporter.java (hunk 3 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/importer/NoteImporter.java:380]: On the opposite. If for some reason the note generate no card, we still want to allow it to be imported and saved in the collection, so `allowEmpty` should be true.

**Comment by Arthur-Milchior** [comment]
Conflict resolved. I also rewrote it to hopefully make each commit simpler to understand, and also add the option to consider cloze and standard note type differently as it was before this PR.

File: AnkiDroid/src/main/java/com/ichi2/libanki/importer/NoteImporter.java (hunk 3 above)
Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/libanki/importer/NoteImporter.java:383]: `ords.isEmpty()` surely should always return false now, unless I'm misunderstanding the code.

Cloze should always generate a card, now a normal note should also generate ord 0?

Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/libanki/importer/NoteImporter.java:383]: @Arthur-Milchior / @david-allison-1 is this point resolved? It seems like this is close, I'd still really like to land it :-)

**Comment by david-allison-1** [comment]
This appears to have broken CI:

```
com.ichi2.anki.tests.ContentProviderTest > testMoveCardsToOtherDeck[0][TestingAVD(AVD) - 11] FAILED 
	android.database.sqlite.SQLiteException: error while compiling: "SELECT id as _id, guid, mid, allow_empty, mod, usn, tags, flds, sfld, csum, flags, data FROM notes WHERE id in (1616393828073,1616393828277,1616393828506,1616393828883,1616393829171,1616393829514,1616393829727,1616393830070)": DBError { info: "SqliteFailure(Error { code: Unknown, extended_code: 1 }, Some(\"no such column: allow_empty\"))", kind: Other }
	at net.ankiweb.rsdroid.database.StreamingProtobufSQLiteCursor.<init>(StreamingProtobufSQLiteCursor.java:57)

com.ichi2.anki.tests.ContentProviderTest > testQueryNoteIds[0][TestingAVD(AVD) - 11] FAILED 
	android.database.sqlite.SQLiteException: error while compiling: "SELECT id as _id, guid, mid, allow_empty, mod, usn, tags, flds, sfld, csum, flags, data FROM notes WHERE id in (1616393842541,1616393843186,1616393843458,1616393843638,1616393843858,1616393844115,1616393844324,1616393844822)": DBError { info: "SqliteFailure(Error { code: Unknown, extended_code: 1 }, Some(\"no such column: allow_empty\"))", kind: Other }
	at net.ankiweb.rsdroid.database.StreamingProtobufSQLiteCursor.<init>(StreamingProtobufSQLiteCursor.java:57)

com.ichi2.anki.tests.ContentProviderTest > testUpdateNoteFields[0][TestingAVD(AVD) - 11] FAILED 
	android.database.sqlite.SQLiteException: error while compiling: "SELECT id as _id, guid, mid, allow_empty, mod, usn, tags, flds, sfld, csum, flags, data FROM notes WHERE id=?": DBError { info: "SqliteFailure(Error { code: Unknown, extended_code: 1 }, Some(\"no such column: allow_empty\"))", kind: Other }
	at net.ankiweb.rsdroid.database.StreamingProtobufSQLiteCursor.<init>(StreamingProtobufSQLiteCursor.java:57)

com.ichi2.anki.tests.ContentProviderTest > testQueryNotesProjection[0][TestingAVD(AVD) - 11] FAILED 
	android.database.sqlite.SQLiteException: error while compiling: "SELECT guid, mid, allow_empty, mod, usn, tags, flds, sfld, csum, flags, data FROM notes WHERE id in (1616393878268,1616393878819,1616393879184,1616393879491,1616393879961,1616393880460,1616393880768,1616393880969)": DBError { info: "SqliteFailure(Error { code: Unknown, extended_code: 1 }, Some(\"no such column: allow_empty\"))", kind: Other }
	at net.ankiweb.rsdroid.database.StreamingProtobufSQLiteCursor.<init>(StreamingProtobufSQLiteCursor.java:57)

com.ichi2.anki.tests.ContentProviderTest > testQueryDirectSqlQuery[0][TestingAVD(AVD) - 11] FAILED 
	android.database.sqlite.SQLiteException: error while compiling: "SELECT id as _id, guid, mid, allow_empty, mod, usn, tags, flds, sfld, csum, flags, data FROM notes WHERE mid=1616393888238": DBError { info: "SqliteFailure(Error { code: Unknown, extended_code: 1 }, Some(\"no such column: allow_empty\"))", kind: Other }
	at net.ankiweb.rsdroid.database.StreamingProtobufSQLiteCursor.<init>(StreamingProtobufSQLiteCursor.java:57)

com.ichi2.anki.tests.ContentProviderTest > testMoveCardsToOtherDeck[1][TestingAVD(AVD) - 11] FAILED 
	android.database.sqlite.SQLiteException: error while compiling: "SELECT id as _id, guid, mid, allow_empty, mod, usn, tags, flds, sfld, csum, flags, data FROM notes WHERE id in (1616393956534,1616393956767,1616393956995,1616393957134,1616393957271,1616393957625,1616393957917,1616393958127)": DBError { info: "SqliteFailure(Error { code: Unknown, extended_code: 1 }, Some(\"no such column: allow_empty\"))", kind: Other }
	at net.ankiweb.rsdroid.database.StreamingProtobufSQLiteCursor.<init>(StreamingProtobufSQLiteCursor.java:57)

com.ichi2.anki.tests.ContentProviderTest > testQueryNoteIds[1][TestingAVD(AVD) - 11] FAILED 
	android.database.sqlite.SQLiteException: error while compiling: "SELECT id as _id, guid, mid, allow_empty, mod, usn, tags, flds, sfld, csum, flags, data FROM notes WHERE id in (1616393967759,1616393967957,1616393968257,1616393968552,1616393969115,1616393969404,1616393969680,1616393969874)": DBError { info: "SqliteFailure(Error { code: Unknown, extended_code: 1 }, Some(\"no such column: allow_empty\"))", kind: Other }
	at net.ankiweb.rsdroid.database.StreamingProtobufSQLiteCursor.<init>(StreamingProtobufSQLiteCursor.java:57)

com.ichi2.anki.tests.ContentProviderTest > testUpdateNoteFields[1][TestingAVD(AVD) - 11] FAILED 
	android.database.sqlite.SQLiteException: error while compiling: "SELECT id as _id, guid, mid, allow_empty, mod, usn, tags, flds, sfld, csum, flags, data FROM notes WHERE id=?": DBError { info: "SqliteFailure(Error { code: Unknown, extended_code: 1 }, Some(\"no such column: allow_empty\"))", kind: Other }
	at net.ankiweb.rsdroid.database.StreamingProtobufSQLiteCursor.<init>(StreamingProtobufSQLiteCursor.java:57)

com.ichi2.anki.tests.ContentProviderTest > testQueryNotesProjection[1][TestingAVD(AVD) - 11] FAILED 
	android.database.sqlite.SQLiteException: error while compiling: "SELECT guid, mid, allow_empty, mod, usn, tags, flds, sfld, csum, flags, data FROM notes WHERE id in (1616394000267,1616394000643,1616394002353,1616394002709,1616394003437,1616394003955,1616394004681,1616394005692)": DBError { info: "SqliteFailure(Error { code: Unknown, extended_code: 1 }, Some(\"no such column: allow_empty\"))", kind: Other }
	at net.ankiweb.rsdroid.database.StreamingProtobufSQLiteCursor.<init>(StreamingProtobufSQLiteCursor.java:57)

com.ichi2.anki.tests.ContentProviderTest > testQueryDirectSqlQuery[1][TestingAVD(AVD) - 11] FAILED 
	android.database.sqlite.SQLiteException: error while compiling: "SELECT id as _id, guid, mid, allow_empty, mod, usn, tags, flds, sfld, csum, flags, data FROM notes WHERE mid=1616394016904": DBError { info: "SqliteFailure(Error { code: Unknown, extended_code: 1 }, Some(\"no such column: allow_empty\"))", kind: Other }
	at net.ankiweb.rsdroid.database.StreamingProtobufSQLiteCursor.<init>(StreamingProtobufSQLiteCursor.java:57)
```
